qiskit/providers/aer/backends/aerbackend.py

```diff
@@ -379,24 +412,35 @@ def _assemble_noise_model(self, circuits, **run_options):
         noise_model = run_options.get(
             'noise_model', getattr(self.options, 'noise_model', None))
 
-        # Add custom pass noise only to QuantumCircuit objects
-        if noise_model and all(isinstance(circ, QuantumCircuit) for circ in circuits):
+        # Add custom pass noise only to QuantumCircuit objects that contain delay
+        # instructions since this is the only instruction handled by the noise pass
+        # at present
+        if noise_model and all(isinstance(circ, QuantumCircuit) for circ in run_circuits):
             npm = noise_model._pass_manager()
             if npm is not None:
-                circuits = npm.run(circuits)
-                if isinstance(circuits, QuantumCircuit):
-                    circuits = [circuits]
+                # Get indicies of circuits that need noise transpiling
+                transpile_idxs = [
+                    idx for idx, opset in enumerate(optypes) if Delay in opset
+                ]
+
+                # Transpile only the required circuits
+                transpiled_circuits = npm.run([run_circuits[i] for i in transpile_idxs])
+                if isinstance(transpiled_circuits, QuantumCircuit):
+                    transpiled_circuits = [transpiled_circuits]
+
+                # Update the circuits with transpiled ones
+                for idx, circ in zip(transpile_idxs, transpiled_circuits):
+                    run_circuits[idx] = circ
+                    optypes[idx] = self._circuit_optypes(circ)
 
         # Check if circuits contain quantum error instructions
-        run_circuits = []
-        for circ in circuits:
-            if isinstance(circ, (Schedule, ScheduleBlock)):
-                run_circuits.append(circ)
-            else:
+        for idx, circ in enumerate(run_circuits):
+            if QuantumChannelInstruction in optypes[idx] and not isinstance(
+                    circ, (Schedule, ScheduleBlock)):
                 updated_circ = False
                 new_data = []
                 for inst, qargs, cargs in circ.data:
-                    if inst.name == "quantum_channel":
+                    if isinstance(inst, QuantumChannelInstruction):
```

Comment:
I don't know if `isinstance` is what you mean here.  It's usually correct in Python, but your `optypes` set is of _exact_ classes, so you'll only enter this loop if there exists an instruction whose type is _exactly_ `QuantumChannelInstruction`.  It's a general problem with storing instructions in the set.

Perhaps my issue here is wiuth `optypes`: instead of just adding `type(instruction)` to the set, you might need to add all types in `type(instruction).__bases__` that are subclasses of `Instruction`/whatever you care about.

Reply:
Is instance should be correct here now that we've fixed the MRO stuff in optypes